### DISCUSSION TITLE
Model calibration - reliability diagram / confidence of predicted labels vs true labels.

### DISCUSSION THREAD

**Discussion by pe-hy** [autogluon/autogluon#3328] [Q&A]
Hi. I'm using TabularPredictor to train a multiclass model (there are about 20+ classes). I would like to look at how the model is calibrated by either plotting reliability diagram or at least obtaining a table such as this:

![image](https://github.com/autogluon/autogluon/assets/70585331/1b33e728-ee18-4987-8ae2-fa3cf4ffc3e9)

How would one approach this?

**Comment by Innixma** [comment]
Could you give us a link to the kind of diagram you are interested in?

**Reply by Innixma** [reply]
To get the table:

```
pred_label = predictor.predict(test_data)
confidences = predictor.predict_proba(test_data)
true_label = test_data[predictor.label]
```

Then do some logic to get the confidence of the pred_label class and rename it `confidence`, then concat.

**Reply by pe-hy** [reply]
Hi Innixma, thank you for the answer. I worked on my problem and finally got kinda close by pretty much "acting" like it's a binary classification, here's some of the code (this all happens after training the model using autogluon):

```
y_test = test_data[label]
pred_probs = predictor.predict_proba(test_data_nolabel)
result_dict = {}

for index, row in pred_probs.iterrows():
    actual_label = test_data.iloc[index, -1]
    highest_prob = row.max()
    predicted_label = row.idxmax()

    if index not in result_dict:
        result_dict[index] = {}

    result_dict[index][actual_label] = (predicted_label, highest_prob)

```
My result_dict looks like this:
```
{0: {5: (5, 0.9710572361946106)},
 1: {7: (7, 0.6950807571411133)},
 2: {8: (8, 0.4435451626777649)}, ...} 
```

Where the outer key is the row id, inner key is the true label, first value in tuple is the predicted label and the second value is the probability.

I then did this (using calibration_curve method from sklearn):

```
y_prob = []
y_true = []
for example_id, in_d in result_dict.items():
    for actual_label, tup in in_d.items():
        pred_label, pred_prob = tup
        if actual_label == pred_label:
            y_prob.append(pred_prob)
            y_true.append(1)
        else:
            y_prob.append(pred_prob)
            y_true.append(0)

# Compute the calibration curve
prob_true, prob_pred = calibration_curve(y_true, y_prob, n_bins=7)

# Plot the calibration curve
plt.plot(prob_pred, prob_true, marker='o', linestyle='-', label='Calibration Curve')
plt.plot([0, 1], [0, 1], linestyle='--', label='Ideal Calibration')
plt.xlabel('Predicted Probability')
plt.ylabel('True Fraction of Positive Outcomes')
plt.legend()
plt.show()
```

To obtain the following diagram:
![image](https://github.com/autogluon/autogluon/assets/70585331/2c30a30b-1623-459e-80fc-9ba8e4edc905)

Although this result seems pretty good, I'm not entirely sure I did this correctly from statistics standpoint.

The result I was going for is a plot kind of like this:

![image](https://github.com/autogluon/autogluon/assets/70585331/6217c22d-d73d-4630-b785-81f08e238f78)
